master/internal/api_checkpoint.go

```diff
@@ -197,24 +167,169 @@ func (a *apiServer) DeleteCheckpoints(
+			"this subset of checkpoints provided are in the model registry and cannot be deleted: %v.",
+			registeredCheckpointUUIDs)
+	}
+
+	err = db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		var updatedCheckpointSizes []uuid.UUID
+		for i, c := range req.Checkpoints {
+			if c.Resources != nil {
+				size := int64(0)
+				for _, v := range c.Resources.Resources {
+					size += v
+				}
+
+				v1Update := tx.NewUpdate().Model(&model.CheckpointV1{}).
+					Where("uuid = ?", c.Uuid)
+				v2Update := tx.NewUpdate().Model(&model.CheckpointV2{}).
+					Where("uuid = ?", c.Uuid)
+
+				if len(c.Resources.Resources) == 0 { // Full delete case.
+					v1Update = v1Update.Set("state = ?", model.DeletedState)
+					v2Update = v2Update.Set("state = ?", model.DeletedState)
+				} else { // Partial delete case.
+					v1Update = v1Update.
+						Set(`state =
+							(CASE WHEN ? @> resources AND resources @> ? THEN state ELSE ? END)`,
+							c.Resources.Resources, c.Resources.Resources, model.PartiallyDeletedState).
+						Set("resources = ?", c.Resources.Resources).
+						Set("size = ?", size)
+					v2Update = v2Update.
+						Set(`state =
+							(CASE WHEN ? @> resources AND resources @> ? THEN state ELSE ? END)`,
+							c.Resources.Resources, c.Resources.Resources, model.PartiallyDeletedState).
+						Set("resources = ?", c.Resources.Resources).
+						Set("size = ?", size)
+				}
+
+				if _, err := v1Update.Exec(ctx); err != nil {
+					return fmt.Errorf("deleting checkpoints from raw_checkpoints: %w", err)
+				}
+
+				if _, err := v2Update.Exec(ctx); err != nil {
+					return fmt.Errorf("deleting checkpoints from checkpoints_v2: %w", err)
+				}
+
+				updatedCheckpointSizes = append(updatedCheckpointSizes, uuids[i])
+			}
+		}
+
+		if len(updatedCheckpointSizes) > 0 {
+			if err := db.UpdateCheckpointSizeTx(ctx, tx, updatedCheckpointSizes); err != nil {
+				return fmt.Errorf("updating checkpoint size: %w", err)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error patching checkpoints: %w", err)
+	}
+
+	return &apiv1.PatchCheckpointsResponse{}, nil
+}
+
+func (a *apiServer) CheckpointsRemoveFiles(
+	ctx context.Context,
+	req *apiv1.CheckpointsRemoveFilesRequest,
+) (*apiv1.CheckpointsRemoveFilesResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	conv := &protoconverter.ProtoConverter{}
+	checkpointsToDelete := conv.ToUUIDList(req.CheckpointUuids)
+	if cErr := conv.Error(); cErr != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "converting checkpoint: %s", cErr)
+	}
+
+	for _, g := range req.CheckpointGlobs {
+		if len(g) == 0 {
+			// Avoid weirdness where someone passes in "" then we concat {uuid}/{glob}
+			// and we unexpectedly delete the whole checkpoint folder.
+			return nil, status.Errorf(codes.InvalidArgument, "cannot have empty string glob")
+		}
+		if strings.Contains(g, "..") {
```

Comment:
is this sufficient to make sure no one passes a glob that could delete all of shared_fs?

Reply:
I can't think of other checks we could do here in go...

however in python, we could check `os.path.common_path(checkpoint_dir, normpath(globbed_path)) == checkpoint_dir` for extra safety.

another wild idea (and a lot of work) is to spawn a python worker per checkpoint uuid, and have each worker `os.chroot` into their respective uuid directory...

Reply:
Good idea on common path check, added it.